### PR TITLE
gh-103256: Fix hmac algorithm to support fallback implementation

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-103256

### CODE REVIEW

**Comment by corona10** [comment]
@gpshead 

Would you like to take a look at this PR? 
The change will not affect to backward compatibility.

File: Modules/_hashopenssl.c [hunk 1]
```diff
@@ -355,7 +355,7 @@ py_digest_by_name(PyObject *module, const char *name, enum Py_hash_type py_ht)
         }
     }
     if (digest == NULL) {
-        _setException(PyExc_ValueError, "unsupported hash type %s", name);
+        _setException(state->unsupported_digestmod_error, "unsupported hash type %s", name);
```
Comment (gpshead) [inline Modules/_hashopenssl.c:358]: nice.  looking around, as this is already a ValueError subclass, this is indeed the ideal fix for this problem. :)

**Review by gpshead — APPROVED** [review]
Thanks!  somewhat surprised we hadn't run into this at work, but I guess there isn't much demand for hmac-sha3 yet.  the other folks who might hit this probably have openssl builds without md5 or sha1?

**Comment by corona10** [comment]
> I guess there isn't much demand for hmac-sha3 yet. the other folks who might hit this probably have openssl builds without md5 or sha1?

It looks so, and the latter looks minor case in this era :)

**Comment by corona10** [comment]
Thanks @gpshead for reviewing and supplementing this PR!

**Comment by geitda** [comment]
> Thanks! somewhat surprised we hadn't run into this at work, but I guess there isn't much demand for hmac-sha3 yet. the other folks who might hit this probably have openssl builds without md5 or sha1?

SHA3 is fine since OpenSSL (at least in Python 3.11.2) provides it;
```
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)] on win32
>>> import hmac
>>> hmac.digest(b'key', b'message', 'sha3_256').hex()
'0f43852a24d5597a8200312a95993991581679d63264f1b1ad4b5ccac7fe8ba4'
```
The only hash algorithms that were a problem are ones that OpenSSL doesn't provide via named lookup. So none of the built-in algorithms would trigger it as OpenSSL provides _all of them_ (even whirlpool, mdc2, and the composite md5-sha1 work as expected) - that's why it took me (ab)using hashlib to find it. But patching hashlib (directly in the file or on the fly) to include a new algorithm does trigger it (as that's what I was doing). But you're right that if md5 or sha1 _were_ removed from OpenSSL while still being kept in hashlib (e.g. with a stand-alone CPython implementation), then the problem _would_ get triggered.

**Comment by gpshead** [comment]
Not everyone's OpenSSL provides it. 1.1.1 and BoringSSL do not.

**Pull request merged**